sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -672,4 +689,300 @@ private WindowDoFnAdapter(DoFnReflector reflector, DoFnWithContext<InputT, Outpu
       super(reflector, fn);
     }
   }
+
+  /**
+   * A byte-buddy {@link Implementation} that delegates a call that receives
+   * {@link ExtraContextInfo} to the given {@link DoFnWithContext} method.
+   */
+  private static final class InvokerDelegation implements Implementation {
+    @Nullable
+    private final Method target;
+    private final boolean isStartBundle;
+    private final ExtraContextInfo[] args;
+    private final Assigner assigner = Assigner.DEFAULT;
+    private FieldDescription field;
+
+    /**
+     * Create the {@link InvokerDelegation} for the specified method.
+     *
+     * @param target the method to delegate to
+     * @param isStartBundle whether or not this is the {@code startBundle} call
+     * @param args the {@link ExtraContextInfo} to be passed to the {@code target}
+     */
+    private InvokerDelegation(
+        @Nullable Method target, boolean isStartBundle, ExtraContextInfo[] args) {
+      this.target = target;
+      this.isStartBundle = isStartBundle;
+      this.args = args;
+    }
+
+    /**
+     * Generate the {@link Implementation} of one of the life-cycle methods of a
+     * {@link DoFnWithContext}.
+     */
+    private static Implementation create(
+        @Nullable final Method target, boolean isStartBundle, ExtraContextInfo[] args) {
+      if (target == null && !isStartBundle) {
+        // There is no target to call, and this is not a startBundle method (which needs to call
+        // prepareForProcessing no matter whaht). There is nothing to do, so just produce a stub:
+        return StubMethod.INSTANCE;
```

Comment:
Added an enum for the before-delegation call (in this case, prepareForProcessing). It could probably be simplified and/or generalized more, but I think this strikes a good trade-off. Its not as specific as the startBundle boolean, documents the logic more clearly, and provides a clear extension point in case more is necessary in the future.